compiler/src/dotty/tools/dotc/ast/Desugar.scala

```diff
@@ -1022,6 +1022,36 @@ object desugar {
     else Apply(ref(tupleTypeRef.classSymbol.companionModule.termRef), ts)
   }
 
+  /** Group all definitions that can't be at the toplebel in
+   *  an object named `<source>#object` where `<source>` is the name of the source file.
+   *  Definitions that can't be at the toplevel are:
+   *
+   *   - all pattern, value and method definitions
+   *   - non-class type definitions
+   *   - implicit classes and objects
+   *   - companion objects of opaque types
+   */
+  def packageDef(pdef: PackageDef)(implicit ctx: Context): PackageDef = {
+    val opaqueNames = pdef.stats.collect {
+      case stat: TypeDef if stat.mods.is(Opaque) => stat.name
+    }
+    def needsObject(stat: Tree) = stat match {
+      case _: ValDef | _: PatDef | _: DefDef => true
+      case stat: ModuleDef =>
+        stat.mods.is(Implicit) || opaqueNames.contains(stat.name.stripModuleClassSuffix.toTypeName)
+      case stat: TypeDef => !stat.isClassDef || stat.mods.is(Implicit)
+      case _ => false
+    }
+    val (nestedStats, topStats) = pdef.stats.partition(needsObject)
+    if (nestedStats.isEmpty) pdef
+    else {
+      val sourceName = ctx.source.file.name.takeWhile(_ != '.')
+      val groupName = (sourceName ++ str.TOPLEVEL_SUFFIX).toTermName
```

Comment:
I'd suggest using a SuffixNameKind to add TOPLEVEL_SUFFIX for this to be able to recover the original file name if needed and for use in `isPackageObject`.

Reply:
I did not follow this path since the name is user-visible (i.e. it is specced that way). Only names that are inaccessible for users are handled by NameKind. E.g. ModuleClassName expands to a trailing `$`, but that's implementation, not spec.